### ISSUE TITLE
About Figure 2 in paper

### ISSUE THREAD

**Issue by Lukangkang123** [jingweio/DSF#1]
Thanks for your great work. I would like to ask, what is the meaning of the vertical and horizontal coordinates in Figure 2 in your paper?As follows）Can you provide the code to draw Figure 2?  Thanks !
![image](https://github.com/jingweio/DSF/assets/38984189/5af6af7e-3337-4c4f-8816-95ccbfd12e92)

**Comment by jingweio** [comment]
Takes Figure (b) as a example, which depicts a histogram. The x-axis shows local graph frequencies derived from the eigenvectors associated with the median eigenvalue and a 2-hop node neighborhood. The y-axis marks the frequency of occurrences within each interval. More details are available in the concluding paragraphs of Section 4.1. The related code is provided below and I also upload the texas dataset as an example. Please feel free to contact me should you encounter any bugs.

```python
from torch_geometric.utils import to_undirected, remove_self_loops
from dataset import load_nc_dataset
import matplotlib.pyplot as plt
import seaborn as sns
import torch
import dgl


def cut_data(x, max_v=1):
    # remove extreme values for better visualization
    return x[x <= max_v]


def plot_hist(data, datname, sav_filename, color_name="#DA382A", title_fontsize=30, ylabel_fontsize=22, tick_fontsize=16, bin_number=30):
    sns.set(style="darkgrid")
    fig, ax = plt.subplots(figsize=(4, 3))
    sns.histplot(data=data, color=color_name, ax=ax, bins=bin_number)
    ax.set_title(datname, {"size": title_fontsize})
    ax.set_ylabel(' ', fontsize=1, color="w")
    ax.set_ylabel('Frequency', fontsize=ylabel_fontsize, color="black")
    ax.ticklabel_format(axis="both", style="sci", scilimits=(0, 0))
    for label in (ax.get_xticklabels() + ax.get_yticklabels()):
        label.set_fontsize(tick_fontsize)
    fig.tight_layout(pad=0.1)
    plt.savefig(sav_filename)
    plt.cla()


def load_dgl_graph(datname, sub_datname=""):
    # load data on dgl-graph
    dataset = load_nc_dataset(datname, sub_datname)
    edge_index = to_undirected(remove_self_loops(dataset.graph["edge_index"])[0])
    G = dgl.graph((edge_index[0], edge_index[1]))
    G.ndata["deg"] = G.in_degrees().float().unsqueeze(dim=-1)
    G.ndata["label"] = dataset.label.unsqueeze(dim=-1)

    # build laplacian
    node_size = dataset.graph['num_nodes']
    A = torch.zeros((node_size, node_size))
    A[edge_index[0], edge_index[1]] = 1
    # exception for not-connected nodes
    valid_idx = (A.sum(dim=1) != 0)
    if not torch.all(valid_idx):
        A = A[valid_idx].T[valid_idx]
        node_size = A.shape[0]
        G.remove_nodes(torch.where(~valid_idx)[0])
    D = torch.sum(A, dim=1, keepdim=True).pow(-0.5)
    A_sym = (D * A).T * D
    L_sym = torch.eye(node_size).float() - A_sym

    # eigen-decomposition
    W, U = torch.linalg.eigh(L_sym)
    Wtopk = torch.topk(W, k=node_size)
    middle_indices = Wtopk.indices[node_size // 2]  # index of the middle-ranked eigenvalues for representative visualization
    G.ndata["U"] = U.T[middle_indices].T
    return G


def U_freq(G):
    edge_index = G.edges()
    U_norm = G.ndata["U"] * (G.ndata["deg"].squeeze().pow(-0.5))
    loc_frq = (U_norm[edge_index[0]] - U_norm[edge_index[1]]).pow(2).sum() / 2
    return loc_frq


def div_local_pattern(G, hops=2):
    result = []
    for node_id in G.nodes():
        print(f"progress: {node_id}/{len(G.nodes())}")
        khop_sg, _ = dgl.khop_out_subgraph(G, node_id, k=hops)
        tmp = U_freq(khop_sg)
        result.append(torch.tensor(tmp))
    result = torch.stack(result, dim=-1)
    return result


def main():
    G = load_dgl_graph("texas")  # load data
    loc_graphFreq_ls = div_local_pattern(G)  # local graph frequency
    plot_hist(data=loc_graphFreq_ls, datname="texas", sav_filename="texas_localGraphFreqHist.png")  # visualization


if __name__ == '__main__':
    main()

**Comment by Lukangkang123** [comment]
Thanks a lot, I have successfully run this code. A small suggestion: You can provide a requirement.txt file in this project so that everyone can reproduce it better.

**Comment by jingweio** [comment]
thanks for the suggestion~
